data/tutorials/lg_08_error_handling.md

```diff
@@ -125,67 +193,569 @@ let () =
+
+The map functions of the `Result` module allows processing of data or errors,
+but the routines used must not fail, as `Result.map` will never turn an `Ok`
+into an `Error` and `Result.map_error` will never turn an `Error` into an `Ok`.
+On the other hand, functions passed to `Result.bind` are allowed to fail. As
+stated before there isn't a `Result.bind_error`. One way to make sense out of
+that absence is to consider its type, it would have to be:
+```ocaml
+val Result.bind_error : ('a, 'e) result -> ('e -> ('a, 'f) result) -> ('a, 'f) result
+```
+We would have:
+* `Result.map_error f (Ok x) = Ok x`
+* And either:
+  - `Result.map_error f (Error e) = Ok y`
+  - `Result.map_error f (Error e) = Error e'`
+
+This means an error would be turned back into valid data or changed into
+another error. This is almost like recovering from an error. However, when
+recovery fails, it may be preferable to preserve the initial cause of failure.
+That behaviour can be achieved by defining the following function:
+
+```ocaml
+# let recover f = Result.(fold ~ok:ok ~error:(fun (e : 'e) -> Option.to_result ~none:e (f e)));;
+val recover : ('e -> 'a option) -> ('a, 'e) result -> ('a, 'e) result = <fun>
+```
+
+Although any kind of data can be wrapped as a `result` `Error`, it is
+recommended to use that constructor to carry actual errors, for instance:
+- `exn`, in which case the result type just makes exceptions explicit
+- `string`, where the error case is a message that indicates what failed
 - `string Lazy.t`, a more elaborate form of error message that is only evaluated
-  if printing is required.
-- some polymorphic variant, with one case per
-  possible error. This is very accurate (each error can be dealt with
-  explicitly and occurs in the type) but the use of polymorphic variants
-  sometimes make error messages hard to read.
+  if printing is required
+- some polymorphic variant, with one case per possible error. This is very
+  accurate (each error can be dealt with explicitly and occurs in the type), but
+  the use of polymorphic variants sometimes make the code harder to read.
+
+Note that some say the types `result` and `Either.t` are
+[ismorphic](https://en.wikipedia.org/wiki/Isomorphism). Concretely, it means
+it's always possible to replace one by the other, like in a completely neutral
+refactoring. Values of type `result` and `Either.t` can be translated back and
+forth, and appling both translations one after the other, in any order, returns
+to the starting value. Nevertheless, this doesn't mean `result` should be used
+in place of `Either.t`, or vice versa. Naming things matters, as punned by Phil
+Karlton's famous quote:
 
-For easy combination of functions that can fail, many alternative standard
-libraries provide useful combinators on the `result` type: `map`, `>>=`, etc.
+> There are only two hard things in Computer Science: cache invalidation and
+> naming things.
```

Comment:
```suggestion
> Two confusing things in Computer Science: cache invalidation and
> naming things.
```
Suggestion as I'm sure there are many hard thing in Computer Science or everyone would be a programmer making the big bucks! 😆  ... but don't feel the need to change it if you like it the way it is. I was taught not to put things like "it's so easy" or "this isn't at all difficult," because if it's difficult for the reader, then they might feel stupid and get discouraged.